src/Workspaces/Core/Portable/Execution/SerializerService_Encoding.cs

```diff
@@ -17,100 +13,103 @@ namespace Microsoft.CodeAnalysis.Serialization
+                    return (encoding.Equals(Encoding.BigEndianUnicode) || HasPreamble(encoding)) ? EncodingId.Unicode_BE_BOM : EncodingId.Unicode_BE;
 
-                return value;
-            }
-            catch (SerializationException)
-            {
-                // even though Encoding is supposed to be serializable, 
-                // not every Encoding follows the rule strictly.
-                // in such as, behave like there was no encoding
-                return null;
-            }
-        }
+                case 12000:
+                    Debug.Assert(HasPreamble(Encoding.UTF32));
+                    return (encoding.Equals(Encoding.UTF32) || HasPreamble(encoding)) ? EncodingId.UTF32_LE_BOM : EncodingId.UTF32_LE;
 
-        public static Encoding? ReadEncodingFrom(ObjectReader reader, CancellationToken cancellationToken)
-        {
-            cancellationToken.ThrowIfCancellationRequested();
+                case 12001:
+                    Debug.Assert(HasPreamble(Encoding.UTF32));
+                    return (encoding.Equals(Encoding.UTF32) || HasPreamble(encoding)) ? EncodingId.UTF32_BE_BOM : EncodingId.UTF32_BE;
 
-            var serialized = reader.ReadByte();
-            if (serialized == EncodingSerialization)
-            {
-                var array = (byte[])reader.ReadValue();
-                var formatter = new BinaryFormatter();
+                case 65001:
+                    Debug.Assert(HasPreamble(Encoding.UTF8));
+                    return (encoding.Equals(Encoding.UTF8) || HasPreamble(encoding)) ? EncodingId.UTF8_BOM : EncodingId.UTF8;
 
-                return (Encoding)formatter.Deserialize(new MemoryStream(array));
+                default:
+                    return EncodingId.Named;
             }
-
-            return ReadEncodingFrom(serialized, reader, cancellationToken);
         }
 
-        private static Encoding? ReadEncodingFrom(byte serialized, ObjectReader reader, CancellationToken cancellationToken)
+        private static bool HasPreamble(Encoding encoding)
+#if NETCOREAPP
+            => !encoding.Preamble.IsEmpty;
+#else
+            => !encoding.GetPreamble().IsEmpty();
+#endif
+
+        public static Encoding? ReadEncodingFrom(ObjectReader reader, CancellationToken cancellationToken)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
-            if (serialized != NoEncodingSerialization)
+            var kind = reader.ReadByte();
+            return ((EncodingId)kind) switch
             {
-                return null;
-            }
-
-            var webName = reader.ReadString();
-            return webName == null ? null : Encoding.GetEncoding(webName);
+                EncodingId.None => null,
+                EncodingId.Named => Encoding.GetEncoding(reader.ReadString()),
+                EncodingId.UTF8 => _cachedEncodings[kind] ??= new UTF8Encoding(encoderShouldEmitUTF8Identifier: false),
```

Comment:
lazily doing this seems like an unneeded micro optimization.

Reply:
You mean we should pre-allocate all these encodings?

Reply:
yup.  i don't have a problem with 4 allocations here.

Reply:
The complexity of that is similar though.